node/core/pvf/src/worker_common.rs

```diff
@@ -199,71 +201,45 @@ where
 }
 
 /// Loop that runs in the CPU time monitor thread on prepare and execute jobs. Continuously wakes up
-/// from sleeping and then either sleeps for the remaining CPU time, or sends back a timeout error
-/// if we exceed the CPU timeout.
-///
-/// NOTE: If the job completes and this thread is still sleeping, it will continue sleeping in the
-/// background. When it wakes, it will see that the flag has been set and return.
-pub async fn cpu_time_monitor_loop(
+/// from sleeping and then either sleeps for the remaining CPU time, or returns if we exceed the CPU
+/// timeout.
+pub fn cpu_time_monitor_loop(
 	job_kind: JobKind,
-	mut stream: UnixStream,
 	cpu_time_start: ProcessTime,
 	timeout: Duration,
-	lock: Arc<AtomicBool>,
+	finished_flag: Arc<AtomicBool>,
```

Comment:
I think actually using https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html would be better here?

Or even better: https://doc.rust-lang.org/stable/std/sync/mpsc/struct.Receiver.html

Or we would make `cpu_time_monitor_loop` async, but that is maybe not perfect for what we want to achieve here. 

Nevertheless, the point I want to bring here is that we should use some kind of `wait_timeout`. Aka we are directly signaled when the other side is finished. The good thing of that is that our thread here would not run for quite some time until the sleep resolves. Depending on the timeout, there could maybe multiple of these "zombie threads" running in the background just waiting for the sleep to finish. (This mainly depends on how fast we reschedule work on this worker).

Reply:
Sorry, I'm not really following. Condvar looks interesting, but it seems to require a mutex? (Haven't used it before.)

I don't think this thread needs to be async. It mostly just waits in a blocking way, but should use basically no resources until it is woken up again. On the other hand if it was async it would be polling or executing async tasks thus consuming CPU time needlessly. But I'm not sure if I understand your point.

Reply:
My point is. The timeout is maybe 6 seconds or more, but the validation is finished after 500ms. So, the thread will wake up 5.5 seconds later. There is no problem with this, but the thread stays alive all the time which isn't really required. So my proposal about changing the way here to abort the blocking the moment the validation is finished. I get that you want to make this async, but then you can also just use a std mpsc receiver here, because that can also recv with timeout.

Reply:
That's a good observation, and it's true that the threads will stay alive longer than necessary. But if they're sleeping they will barely consume CPU, if any, until a hardware interrupt. IOW, a sleeping thread is basically suspended, whereas doing it in an async way will have an active thread and a bit more overhead, right? (Correct me if I'm wrong.) Either way, it's not a big enough difference that I have a strong preference. I asked on a previous PR if we would ever end up with a lot of sleeping threads doing it this way, and I believe @slumber said the jobs are not frequent enough for it to be a problem.

Reply:
It's a good point, and I actually might re-do this. 🙃 Part of my thinking was that the threads should use as little CPU as possible, so that we measure only the work being done. *However*, I just realized that while doing a job, 5 threads from previous jobs might wake up and we would need to process and kill them, consuming CPU. This affects the determinism of jobs.

For huge timeouts we could go with a max sleep time (say, 5s if the timeout is 1min) - but might be better to just go with the cleaner way. :)